audio2tap.1
.TH audio2tap 1 "17th May, 2003"
.SH NAME
audio2tap \- Create .TAP files from Commodore 64 tapes
.SH SYNOPSIS
audio2tap -h|-v
.P
audio2tap [-i] [-f <freq>] [-d <min duration>] [-H <min height>] [-0] [-c <c64|c16|c16semi|vic20>] [-n] [-t <initial_threshold>] <output TAP file> [input audio file]

.SH DESCRIPTION
.I audio2tap
converts Commodore 64 tapes files to .TAP files.
The tape can be either converted directly from the sound card (by default) or
from an audio file (.WAV or similar, if an input file name is
specified as second argument).
\fIlibtapencoder.so\fR is required. Conversion from sound card requires
\fIlibportaudio.so.2\fR. Conversion from WAV file requires \fIlibaudiofile.so.0\fR.
.P
.SH OPTIONS
.TP
.B -h
Show a help message and exit successfully.
.TP
.B -v
Show version and copyright info and exit successfully
.TP
.B -d <min duration>
When decoding a sound to pulses, pulses which last less that <min duration>
samples are ignored. This option is useful for noisy sounds, in order to
avoid false triggers. The default value is 0.
.TP
.B -f <freq>
Sets the sampling frequency of the input sound to <freq> Hz. The default is
44100. This is only significant if the input is from sound card. If the input
is from audio file, this option is ignored and the frequency will be read
from the file itself. As long as the frequency is supported by the sound card,
Modern sound cards easily reach 96000 Hz, therefore use that if yours supports it.
.TP
.B -H <sensitivity>
With sensitivity 0, maximums and minimums are ignored in case they
come before the trigger has fired. There can be cases, however, in which such maximums and minimums
have to be considered. If the sensitivity is > 0 but low, maximums a little below the trigger level
and minimums a little above the trigger lever will be considered as well. As the sensitivity increases,
more maximums and minimums will be considered. If the sensitivity is 100, all maximums and minimums
will be considered. It is only recommended to use high values of sensitivity with bad tapes, in
presence of sudden sound level changes, otherwise beter keep this low. Default is 12.
.TP
.B -i
Tell
.I audio2tap
that the .WAV file or the sound coming from the sound card has an inverted
waveform. Some soundcards invert the signal when sampling, i.e. they turn
positive waveform values into negative ones and vice versa. If the .WAV file
or the sound was generated with such a soundcard,
.I audio2tap
must be run with the
.B -i
option in order to detect the files. If you get no results without the
.B -i
option, try enabling it. The option is ignored if the machine is C16 with semiwaves.
.TP
.B -0
Create a version-0 TAP file. By default, a version-1 TAP file is created.
.TP
.B -c <c64|c16|c16semi|vic20>
Use timing for C64, C16 or VIC20. \fBc16semi\fR also enables semiwave recording.
Default is \fBc64\fR.
.TP
.B -n
use NTSC timing
.TP
.B -t <initial_threshold>
Ignore pulses at the beginning of the file until the amplitude of the signal exceeds
a range centred around \fB<initial_threshold>\fR. Possible values: 0-255. Default: 20.
.SH AUTHOR
Fabrizio Gennari (fabrizio.ge@tiscalinet.it)
.SH REFERENCES
.I http://www.computerbrains.com/tapformat.html
.IP
Per Hakan Sundell's description of .TAP file format
.PP
.I http://www.portaudio.com
.IP
PortAudio library home page. \fIaudio2tap\fR needs it for reading C64 tapes
from sound card
.PP
.I http://www.68k.org/~michael/audiofile
.IP
Audiofile library home page. \fIaudio2tap\fR needs it for reading C64 tapes
from WAV files
.SH SEE ALSO
.B prg2wav
(1),
.B wav2prg
(1),
.B tap2audio
(1)


audio2tap.c
/* Audiotap: a program for playing TAP files or converting them to WAV files,
 * and to convert C64 tapes to TAP files
 *
 * Copyright (c) Fabrizio Gennari, 2003
 *
 * The program is distributed under the GNU General Public License.
 * See file LICENSE.TXT for details.
 *
 * audio2tap.c : main audio->tap program for the command-line version
 * 
 * This file belongs to the audio->tap part
 * This file is part of the command-line version of Audiotap
 */

#include <fcntl.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <signal.h>

#include "audio2tap_core.h"
#include "audiotap.h"
#include "audiotap_loop.h"

static void sig_int(int signum){
  audiotap_interrupt();
}

void help(){
  printf("Usage: audio2tap -h|-V\n");
  printf("       audio2tap [-i] [-d min_duration] [-H sensitivity] [-f <freq>] [-0] [-c c64ntsc|vicpal|vicntsc|c16pal|c16ntsc] <output TAP file> [input WAV file]\n");
  printf("Options:\n");
  printf("\t-h: show this help message and exit successfully\n");
  printf("\t-V: show version and copyright info and exit successfully\n");
  printf("\t-i: use inverted waveforms (ignored if -c c16semi present)\n");
  printf("\t-d <min_duration>: ignore pulses if distance between min and max is less than or equal to <min_duration> samples (default:0)\n");
  printf("\t-H <sensitivity>: set sensitivity to <sensitivity> (range 0-100, default:12)\n");
  printf("\t-f: use input frequency <freq> Hz, default 44100 (only if input is sound card)\n");
  printf("\t-0: Generate a TAP file of version 0\n");
  printf("\t-c <c64|c16|c16semi|vic20>: Set clock frequency to match specified Commodore machine (default c64)\n");
  printf("\t-n: use NTSC timing\n");
  printf("\t-t <initial_threshold>: ignore initial pulses until their amplitude exceeds a range centred around <initial_threshold> (0-255, default:20)\n");
}

void version(){
  printf("audio2tap (part of Audiotap) version 1.5\n");
  printf("(C) by Fabrizio Gennari, 2003-2011\n");
  printf("This program is distributed under the GNU General Public License\n");
  printf("Read the file LICENSE.TXT for details\n");
  printf("This product includes software developed by the NetBSD\n");
  printf("Foundation, Inc. and its contributors\n");
}
   
int main(int argc, char** argv){
  struct audiotap_init_status status;
  struct tapenc_params params = {
  /* min_duration */ 0,
  /* sensitivity */ 12,
  /* initial_threshold */ 20,
  /* inverted */ TAP_TRIGGER_ON_RISING_EDGE
  };
  uint32_t freq = 44100;
  uint8_t tap_version = 1;
  struct option cmdline[]={
    {"help"              ,0,NULL,'h'},
    {"version"           ,0,NULL,'V'},
    {"inverted-waveform" ,0,NULL,'i'},
    {"min-duration"      ,1,NULL,'d'},
    {"min-height"        ,1,NULL,'H'},
    {"freq"              ,1,NULL,'f'},
    {"tap-version-0"     ,0,NULL,'0'},
    {"clock"             ,1,NULL,'c'},
    {"ntsc"              ,0,NULL,'n'},
    {"initial-threshold" ,1,NULL,'t'},
    {NULL                ,0,NULL,0}
  };
  char *infile, *outfile;
  int option;
  enum machines clock = MACHINE_C64;
  uint8_t videotype = TAP_VIDEOTYPE_PAL;

  status = audiotap_initialize();
  if ((status.audiofile_init_status != LIBRARY_OK &&
       status.portaudio_init_status != LIBRARY_OK)
    || status.tapencoder_init_status != LIBRARY_OK){
    printf("Failed to initialize audiotap library: both audiofile and pablio failed to load");
    exit(1);
  }
  
  while( (option=getopt_long(argc,argv,"d:H:0ihVc:nw:t:",cmdline,NULL)) != -1){
    switch(option){
    case 'c':
      if(!strcmp(optarg,"c64"))
        clock = MACHINE_C64;
      else if(!strcmp(optarg,"vic"))
        clock = MACHINE_VIC20;
      else if(!strcmp(optarg,"c16"))
        clock = MACHINE_C16;
      else if(!strcmp(optarg,"c16semi"))
        clock = MACHINE_C16_SEMIWAVES;
      else{
        printf("Wrong argument to option -c\n");
        exit(1);
      }
      break;
    case 'd':
      params.min_duration=atoi(optarg);
      break;
    case 't':
      params.initial_threshold=atoi(optarg);
      break;
    case 'H':
      params.sensitivity=atoi(optarg);
      if (params.sensitivity > 100){
        printf("Wrong argument to option -H, must be in range 0-100\n");
        exit(1);
      }
      break;
    case 'f':
      freq=atoi(optarg);
      break;
    case '0':
      tap_version=0;
      break;
    case 'i':
      params.inverted=TAP_TRIGGER_ON_FALLING_EDGE;
      break;
    case 'n':
      videotype=TAP_VIDEOTYPE_NTSC;
      break;
    case 'h':
      help();
      exit(0);
    case 'V':
      version();
      exit(0);
    default:
      help();
      exit(1);
    }
  }

  argc-=optind;
  argv+=optind;

  if(argc == 0){
    printf("You must specify the output file name!\n");
    exit(1);
  }

  if(strlen(argv[0]) < 5 ||
     (strcmp(argv[0]+strlen(argv[0])-4,".tap") &&
      strcmp(argv[0]+strlen(argv[0])-4,".TAP"))){
    outfile=malloc(strlen(argv[0]+5));
    if (outfile == NULL){
      printf("Not enough memory!\n");
      exit(1);
    }
    sprintf(outfile, "%s.tap", argv[0]);
  }
  else outfile = argv[0];

  if (argc == 1){
    if (status.portaudio_init_status != LIBRARY_OK){
      printf("Cannot read from sound card: pablio library missing or invalid\n");
      exit(1);
    }
    infile=NULL;
  }
  else{
    if (status.audiofile_init_status != LIBRARY_OK){
      printf("Cannot read from file: audiofile library missing or invalid\n");
      exit(1);
    }
    infile=argv[1];
  }      
  signal(SIGINT, sig_int);

  audio2tap(infile, outfile, freq, &params, tap_version, clock, videotype);

  exit(0);
}
